### DISCUSSION TITLE
Getting error with ExpressionEvaluatorFactoryHolder docker

### DISCUSSION THREAD

**Discussion by Sovannak22** [jxlsteam/jxls#276] [Q&A] [answered]
Hello I work with jxls for generating excel from template. It work well on my machine until it get deploy to docker it alway got error "[onPool-worker-2] k.g.n.e.handler.GlobalErrorAttributes    : java.lang.NullPointerException: Cannot invoke "org.jxls.expression.ExpressionEvaluatorFactory.createExpressionEvaluator(String)" because the return value of "org.jxls.util.JxlsHelper$ExpressionEvaluatorFactoryHolder.access$100()" is null" is there any additional configuration I need to add to make this work? Thanks you.

**Comment by SoltauFintel** [comment]
I guess I have never tried it with Docker.

There is a special service loading mechanism. Please check if the file org.jxls.expression.ExpressionEvaluatorFactory is in the Docker container. It should be in the jxls jar, folder META-INF/services.

**Reply by Sovannak22** [reply]
@SoltauFintel I've checked the file is right there.

**Reply by SoltauFintel** [reply]
Try another solution: Make a copy of JxlsHelper.java and write your own `ExpressionEvaluator createExpressionEvaluator(final String expression)` implementation that instantiates a JEXL ExpressionEvaluator class without the ServiceLoader mechanism.

If that doesn't help, you'll have to look for a solution yourself, as I haven't used Docker for JXLS yet and don't know how to create the Docker image. Feel free to tell us your solution later.

**Reply by Sovannak22** [reply]
@SoltauFintel I think to cause of this is at TransformationConfig which at the end will invoke function createExpressionEvaluator from JXlsHelper.java if I implement my copy JxlsHelper how can I make TransformationConfig call my createExpressionEvaluator instead?

**Reply by SoltauFintel** [reply]
Make also a copy of TransformationConfig and change it. And so on.

Or fork this project, edit the sources and build the JAR files.

**Comment by dsmf** [comment]
We have the same issue when we package our app with spring-boot-maven-plugin. All is well if app is started directly in IntelliJ. However if we start the packaged app we get the error.

jdk 17
spring boot version 3.1.3
jxls.version 2.12.0
jxls-poi.version 2.12.0

If we use maven-assembly-plugin with descriptorRef jar-with-dependencies for packaging the app, the error does not occur. But we do not know why this solves the issue.

**Reply by leonate** [reply]
Are you able to create a test or a simple app demonstrating the issue?

**Reply by dsmf** [reply]
Not quickly, its a customer project so I'd have to build something from scratch for sharing.

Also I am not sure what exactly is the problematic place. Seems to be something with evaluating the expressions inside the excel template?
Here is our anonymized stacktrace: 
```
14:12:15.103 | ForkJoinPool.commonPool-worker-7 | ERROR | de.foo.bar.processor.writer.Foo.FooWriter | Method createTransformer of org.jxls.transform.poi.PoiTransformer class thrown an exception:
null |  |
org.jxls.common.JxlsException: Method createTransformer of org.jxls.transform.poi.PoiTransformer class thrown an exception:
null
	at org.jxls.util.TransformerFactory.createTransformer(TransformerFactory.java:50)
	at de.foo.bar.processor.writer.Foo.FooWriter.writeExcelFile(FooWriter.java:216)
	at de.foo.bar.processor.writer.Foo.FooWriter.createFile(FooWriter.java:160)
	at de.foo.bar.processor.writer.Foo.FooWriter.lambda$createFiles$0(FooWriter.java:134)
	at java.base/java.util.stream.IntPipeline$1$1.accept(IntPipeline.java:180)
	at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
	at java.base/java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:711)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:116)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.jxls.util.TransformerFactory.createTransformer(TransformerFactory.java:43)
	... 17 common frames omitted
Caused by: java.lang.NullPointerException: Cannot invoke "org.jxls.expression.ExpressionEvaluatorFactory.createExpressionEvaluator(String)" because the return value of "org.jxls.util.JxlsHelper$ExpressionEvaluatorFactoryHolder.access$100()" is null
	at org.jxls.util.JxlsHelper.createExpressionEvaluator(JxlsHelper.java:316)
	at org.jxls.transform.TransformationConfig.<init>(TransformationConfig.java:17)
	at org.jxls.transform.AbstractTransformer.<init>(AbstractTransformer.java:28)
	at org.jxls.transform.poi.PoiTransformer.<init>(PoiTransformer.java:95)
	at org.jxls.transform.poi.PoiTransformer.<init>(PoiTransformer.java:85)
	at org.jxls.transform.poi.PoiTransformer.<init>(PoiTransformer.java:77)
	at org.jxls.transform.poi.PoiTransformer.createTransformer(PoiTransformer.java:151)
	at org.jxls.transform.poi.PoiTransformer.createTransformer(PoiTransformer.java:142)
	at org.jxls.transform.poi.PoiTransformer.createTransformer(PoiTransformer.java:123)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 19 common frames omitted
```

Is my assumption correct that I should be able to reproduce using any temlate using any expression inside the template? 

If that's the case it should be reproducible by generating a hello world rest webservice that produces a simple excel file via JXLS+POI using an excel template.  
Then package with 

 ```
 <build>
        <plugins>
          <plugin>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-maven-plugin</artifactId>
            <configuration>
              <classifier>spring-boot</classifier>
              <mainClass>${main.class}</mainClass>
            </configuration>
            <executions>
              <execution>
                <goals>
                  <goal>repackage</goal>
                </goals>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
```

Then mvn package. 
After that run the generated executable from commandline. 
curl the hello world rest service.

**Reply by dsmf** [reply]
Tried to reproduce in simple demo
https://github.com/dsmf/jxls_try_reproduce_NPE_ExpressionEvaluatorFactory
but I could not reproduce the situation yet.

There must be something different in our real project, but I don't have any idea what it is at the moment.
The only thing that comes to my mind is that we don't have the XLS comments variant in the real project. We use the java API with an EachCommand and a special CellRefGenerator there. But actually these lines of code come after the line where the error occurs.

**Reply by gastendonk** [reply]
@dsmf Please convert the .xlsm file to an .xlsx file.

**Reply by dsmf** [reply]
@gastendonk file type shouldn't make a difference, except for security reasons, should it? (in our project we have vba code included, omitted that in the demo). Anyway, just changed it to xlsx.

**Reply by gastendonk** [reply]
I didn't write that because of the JXLS functionality, but because I won't open macro files from the Internet (virus protection).

**Comment by gastendonk** [comment]
I guess this problem is Spring Boot specific. The jxls JAR files must be unpacked (=not inside a fat JAR). I don't know how this is done with Maven.

**Reply by dsmf** [reply]
Hi @gastendonk,

why do you think that it has to be unpacked? Is this just a guess? I mean fat JAR works with other stuff flawlessly.

Furthermore, I don't think that that's the problem. In the above mentioned [demo](https://github.com/dsmf/jxls_try_reproduce_NPE_ExpressionEvaluatorFactory) I could not reproduce it even though JXLS is in the fat JAR. Must be something more specific to the way we use JXLS in our project I think. 

But as I said further above our workaround is to package via maven assembly plugin. This is quite simple and how this is done is also shown in the [demo](https://github.com/dsmf/jxls_try_reproduce_NPE_ExpressionEvaluatorFactory), see the other maven profile.

**Reply by gastendonk** [reply]
yes, a guess. We don't have a solution because we can't reproduce it.

> ... maven assembly plugin ... maven profile.

I'm barely familiar with Maven.

I think the quickest solution will be for you to download the jxls sources, replace `INSTANCE = loadService(ExpressionEvaluatorFactory.class);` in JxlsHelper.java with ~ `INSTANCE = new ExpressionEvaluatorFactoryJexlImpl();`, build the JAR files and use them.

**Reply by dsmf** [reply]
Thanks, but as maven assembly works as workaround I don't need another workaround. I want to understand why it happens because it leaves a somewhat bad feeling. I'll try to narrow it down when I find time. Will let you know when I can reproduce it in the demo.

**Comment by Sovannak22** [comment]
I noted the issue is when building .jar file not exactly docker. 
```
dependencies {
	implementation(fileTree(mapOf("dir" to "libs", "include" to listOf("*.jar"))))
} 
```
do you think it could be because of this ?

**Reply by dsmf** [reply]
Hi @Sovannak22, 

as you can see from my comments above I am also sure it is not from docker because we have the very same issue when we build via spring-boot-maven-plugin instead of maven-assembly-plugin. 

Do you also use the spring-boot-maven-plugin? 

Where is the code snippet that you mentioned above taken from?

**Reply by Sovannak22** [reply]
Hello @dsmf it's from spring boot also but we use gradle here it's from gradle.kts file.

**Comment by Sovannak22** [comment]
The real issue is because of async I use reactive when the generating file outside reactive function and the issue is gone I tried with @Async annotation with CompletableFuture also meet this issue.

**Reply by dsmf** [reply]
While this may narrow down the issue, I don't think it is a solution to make everything synchronous. In our project we need to generate a lot of files and therefore we need asynchronous processing. So even though this comment may be helpful, I don't think it should be marked as answer.

**Reply by dsmf** [reply]
I added an async method in my example project, but I could not reproduce. So to me it does not seem to be related to async at all: https://github.com/dsmf/jxls_try_reproduce_NPE_ExpressionEvaluatorFactory/commit/b5071edd0c97640ebd404332900d660c3ca2376d 

@Sovannak22, are you sure running synchronously fixed it for you not only sporadically?

**Reply by dsmf** [reply]
Just added experiment with supplyAsync in service and get() on it in the controller. Here I get the NPE. See hello2 endpoint  here: https://github.com/dsmf/jxls_try_reproduce_NPE_ExpressionEvaluatorFactory/commit/f1f2bd1b279945cb2c0c5ead709678f76f96938e

**Reply by SoltauFintel** [reply]
@dsmf  You could build and use Jxls from branch feature-3.0. Creating the Transformer is different there.

**Reply by SoltauFintel** [reply]
@dsmf 3.0.0-beta1 is released. Please try it out.

**Comment by Volburt7** [comment]
For me the issue occurred in an async context with Spring's @Async annotation and only when running the fat jar packaged with spring-boot-maven-plugin.

I solved it by replacing this:

```
final Transformer transformer = TransformerFactory.createTransformer(is, os);
```

with that:

```
Transformer transformer;
final ClassLoader originalThreadClassloader = Thread.currentThread().getContextClassLoader();
Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
try {
   transformer = TransformerFactory.createTransformer(is, os);
finally {
   Thread.currentThread().setContextClassLoader(originalThreadClassloader
}
```

The `ServiceFactoryDefaultImpl` of jxls-2.12.0 uses `ServiceLoader.load(interfaceClass)` which then uses the thread's context classloader `Thread.currentThread().getContextClassLoader()`, so I am setting it to the "normal" classloader before and then finally back.

I got the idea from here: https://stackoverflow.com/questions/1771679/difference-between-threads-context-class-loader-and-normal-classloader

**Comment by fachel** [comment, marked as answer]
When will this problem be fixed? When will the be release be?

**Reply by gastendonk** [reply]
I think it will be fixed with Jxls 3. The process there is completely different. Version 3.0.0-beta1 will be released in the next few days.

**Reply by gastendonk** [reply]
@fachel 3.0.0-beta1 is released. See also our new homepage with new docs: https://jxls.sourceforge.net

**Reply by fachel** [reply]
@gastendonk thanks 🔥
